src/datanode/src/heartbeat/handler/downgrade_region.rs

```diff
@@ -13,38 +13,399 @@
         Box::pin(async move {
-            match self.region_server.set_readonly_gracefully(region_id).await {
-                Ok(SetReadonlyResponse::Success { last_entry_id }) => {
-                    InstructionReply::DowngradeRegion(DowngradeRegionReply {
-                        last_entry_id,
-                        exists: true,
-                        error: None,
-                    })
+            let Some(writable) = self.region_server.is_writable(region_id) else {
+                return InstructionReply::DowngradeRegion(DowngradeRegionReply {
+                    last_entry_id: None,
+                    exists: false,
+                    error: None,
+                });
+            };
+
+            // Ignores flush request
+            if !writable {
+                return self.set_readonly_gracefully(region_id).await;
+            }
+
+            let region_server_moved = self.region_server.clone();
+            if let Some(flush_timeout) = flush_timeout {
+                let register_result = self
+                    .downgrade_tasks
+                    .try_register(
+                        region_id,
+                        Box::pin(async move {
+                            info!("Flush region: {region_id} before downgrading region");
+                            region_server_moved
+                                .handle_request(
+                                    region_id,
+                                    RegionRequest::Flush(RegionFlushRequest {
+                                        row_group_size: None,
+                                    }),
+                                )
+                                .await?;
+
+                            Ok(())
+                        }),
+                    )
+                    .await;
+
+                if register_result.is_busy() {
+                    warn!("Another flush task is running for the region: {region_id}");
                 }
-                Ok(SetReadonlyResponse::NotFound) => {
-                    InstructionReply::DowngradeRegion(DowngradeRegionReply {
-                        last_entry_id: None,
-                        exists: false,
-                        error: None,
-                    })
+
+                let mut watcher = register_result.into_watcher();
+                let result = self.catchup_tasks.wait(&mut watcher, flush_timeout).await;
+
+                match result {
+                    WaitResult::Timeout => {
+                        InstructionReply::DowngradeRegion(DowngradeRegionReply {
+                            last_entry_id: None,
+                            exists: true,
+                            error: Some(format!(
+                                "Flush region: {region_id} before downgrading region is timeout"
+                            )),
+                        })
+                    }
+                    WaitResult::Finish(Ok(_)) => self.set_readonly_gracefully(region_id).await,
```

Comment:
Why set readonly after the flush? I think it's the opposite, otherwise there could be remaining data in the memory.

Reply:
It's okay If users use the remote WAL, and memory data will be replayed via remote WAL on another datanode.

Reply:
> Then is it doable if set readonly first? I think it can remove the odd codes in line 68~70.

It needs to introduce another state, e.g., Downgrading. The region will reject any insert requests and flush all memory data, then set the region to Readonly.